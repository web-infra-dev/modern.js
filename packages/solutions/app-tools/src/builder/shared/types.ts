import type { BuilderPluginAPI as WebpackBuilderPluginAPI } from '@modern-js/builder-webpack-provider';
import type { BuilderPluginAPI as RspackBuilderPluginAPI } from '@modern-js/builder-rspack-provider';
import { BuilderInstance } from '@modern-js/builder-shared';
import { AppNormalizedConfig, Bundler, IAppContext } from '../../types';

type Parameter<T extends (arg: any) => void> = Parameters<T>[0];
type FnParameter<
  T extends {
    [p: string]: (arg: any) => void;
  },
> = {
  [P in keyof T]: Parameter<T[P]>;
};

type BuilderPluginCallbacks<B extends Bundler> = FnParameter<
  Partial<
    Pick<
      B extends 'rspack' ? RspackBuilderPluginAPI : WebpackBuilderPluginAPI,
      | 'onAfterBuild'
      | 'onAfterCreateCompiler'
      | 'onAfterStartDevServer'
      | 'onBeforeBuild'
      | 'onBeforeCreateCompiler'
      | 'onBeforeStartDevServer'
      | 'onDevCompileDone'
      | 'onExit'
    >
  >
>;

export type BuilderOptions<B extends Bundler> = {
  normalizedConfig: AppNormalizedConfig<B>;
  appContext: IAppContext;
} & BuilderPluginCallbacks<B>;

export type ModifyBuilderConfig<B extends Bundler> = (
  config: AppNormalizedConfig<B>,
) => Promise<void> | void;

export type ModifyBuilderInstance = (
  builder: BuilderInstance,
) => Promise<void> | void;
