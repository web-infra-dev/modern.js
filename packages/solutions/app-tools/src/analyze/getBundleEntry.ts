import path from 'path';
import {
  ensureAbsolutePath,
  fs,
  findExists,
  MAIN_ENTRY_NAME,
} from '@modern-js/utils';
import type { Entrypoint } from '@modern-js/types';
import type { AppNormalizedConfig, IAppContext } from '../types';
import { getFileSystemEntry } from './getFileSystemEntry';
import { JS_EXTENSIONS } from './constants';

const ensureExtensions = (file: string) => {
  if (!path.extname(file)) {
    return findExists(JS_EXTENSIONS.map(ext => `${file}${ext}`)) || file;
  }
  return file;
};

const ifAlreadyExists = (
  entrypoints: Entrypoint[],
  checked: Entrypoint,
): boolean =>
  entrypoints.some(entrypoint => {
    if (
      ensureExtensions(entrypoint.entry) === ensureExtensions(checked.entry)
    ) {
      // reset entryName
      checked.entryName = entrypoint.entryName;
      return true;
    }
    // filesystem routes entrypoint conflict with normal entrypoint.
    if (
      entrypoint.entry.startsWith(checked.entry) ||
      checked.entry.startsWith(entrypoint.entry)
    ) {
      throw new Error(
        `Entry configuration conflicts\n Your configuration: ${checked.entry}.\n Default entrypoint: ${entrypoint.entry}\n Please reset source.entries or set source.disableDefaultEntries to disable the default entry rules.`,
      );
    }

    return false;
  });

export const getBundleEntry = (
  appContext: IAppContext,
  config: AppNormalizedConfig,
) => {
  const { appDirectory, packageName } = appContext;
  const {
    source: { disableDefaultEntries, entries, entriesDir },
  } = config;

  const defaults = disableDefaultEntries
    ? []
    : getFileSystemEntry(appContext, config);

  // merge entrypoints from user config with directory convention.
  if (entries) {
    Object.keys(entries).forEach(name => {
      const value = entries[name];
      const entrypoint =
        typeof value === 'string'
          ? {
              entryName: name,
              entry: ensureAbsolutePath(appDirectory, value),
              isAutoMount: true,
              fileSystemRoutes: fs
                .statSync(ensureAbsolutePath(appDirectory, value))
                .isDirectory()
                ? {}
                : undefined,
            }
          : {
              entryName: name,
              entry: ensureAbsolutePath(appDirectory, value.entry),
              isAutoMount: !value.disableMount,
              fileSystemRoutes: value.enableFileSystemRoutes ? {} : undefined,
            };

      if (!ifAlreadyExists(defaults, entrypoint)) {
        defaults.push(entrypoint);
      }
    });
  }

  if (!disableDefaultEntries) {
    // find main entry point which server route is '/'.
    const entriesDirAbs = ensureAbsolutePath(appDirectory, entriesDir || '');
    const found = defaults.find(
      ({ entryName, entry, nestedRoutesEntry = '' }) =>
        entryName === packageName ||
        path.dirname(entry) === entriesDirAbs ||
        path.dirname(nestedRoutesEntry) === entriesDirAbs,
    );
    found && (found.entryName = MAIN_ENTRY_NAME);
  }

  return defaults;
};
