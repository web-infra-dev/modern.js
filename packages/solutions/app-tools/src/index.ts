import path from 'path';
import { lintPlugin } from '@modern-js/plugin-lint';
import {
  cleanRequireCache,
  emptyDir,
  getCommand,
  getArgv,
  fs,
  NESTED_ROUTE_SPEC_FILE,
} from '@modern-js/utils';
import { CliPlugin } from '@modern-js/core';
import { getLocaleLanguage } from '@modern-js/plugin-i18n/language-detector';
import initializePlugin from './plugins/initialize';
import analyzePlugin from './plugins/analyze';
import serverBuildPlugin from './plugins/serverBuild';
import deployPlugin from './plugins/deploy';
import { AppTools } from './types';
import { hooks } from './hooks';
import { i18n } from './locale';

import { restart } from './utils/restart';
import { generateWatchFiles } from './utils/generateWatchFiles';
import {
  buildCommand,
  deployCommand,
  devCommand,
  inspectCommand,
  newCommand,
  serverCommand,
  upgradeCommand,
} from './commands';

export { dev } from './commands/dev';

export { mergeConfig } from '@modern-js/core';
export * from './defineConfig';
// eslint-disable-next-line import/export
export * from './types';

// eslint-disable-next-line import/export
export type { RuntimeUserConfig } from './types/config';

export type AppToolsOptions = {
  /**
   * Specify which bundler to use for the build.
   * @default `webpack`
   * */
  bundler?: 'experimental-rspack' | 'webpack';
};
/**
 * The core package of the framework, providing CLI commands, build capabilities, configuration parsing and more.
 */
export const appTools = (
  options: AppToolsOptions = {
    bundler: 'webpack',
  },
): CliPlugin<AppTools<'shared'>> => ({
  name: '@modern-js/app-tools',

  post: [
    '@modern-js/plugin-initialize',
    '@modern-js/plugin-analyze',
    '@modern-js/plugin-ssr',
    '@modern-js/plugin-document',
    '@modern-js/plugin-state',
    '@modern-js/plugin-router',
    '@modern-js/plugin-router-v5',
    '@modern-js/plugin-polyfill',
  ],

  registerHook: hooks,

  usePlugins: [
    initializePlugin({
      bundler:
        options?.bundler === 'experimental-rspack' ? 'rspack' : 'webpack',
    }),
    analyzePlugin({
      bundler:
        options?.bundler === 'experimental-rspack' ? 'rspack' : 'webpack',
    }),
    serverBuildPlugin(),
    lintPlugin(),
    deployPlugin(),
  ],

  setup: api => {
    const appContext = api.useAppContext();
    api.setAppContext({
      ...appContext,
      toolsType: 'app-tools',
    });
    const nestedRoutes: Record<string, unknown> = {};

    const locale = getLocaleLanguage();
    i18n.changeLanguage({ locale });

    return {
      async beforeConfig() {
        const userConfig = api.useConfigContext();
        const appContext = api.useAppContext();
        if (userConfig.output?.tempDir) {
          api.setAppContext({
            ...appContext,
            internalDirectory: path.resolve(
              appContext.appDirectory,
              userConfig.output.tempDir,
            ),
          });
        }
      },
      async commands({ program }) {
        await devCommand(program, api);
        await buildCommand(program, api);
        serverCommand(program, api);
        deployCommand(program, api);
        newCommand(program, locale);
        inspectCommand(program, api);
        upgradeCommand(program);
      },

      async prepare() {
        const command = getCommand();
        if (command === 'deploy') {
          const isSkipBuild = ['-s', '--skip-build'].some(tag => {
            return getArgv().includes(tag);
          });
          // if skip build, do not clean dist path
          if (isSkipBuild) {
            return;
          }
        }

        // clean dist path before building
        if (
          command === 'dev' ||
          command === 'start' ||
          command === 'build' ||
          command === 'deploy'
        ) {
          const resolvedConfig = api.useResolvedConfigContext();
          if (resolvedConfig.output.cleanDistPath) {
            const appContext = api.useAppContext();
            await emptyDir(appContext.distDirectory);
          }
        }
      },

      async watchFiles() {
        const appContext = api.useAppContext();
        const config = api.useResolvedConfigContext();
        return await generateWatchFiles(appContext, config.source.configDir);
      },

      // 这里会被 core/initWatcher 监听的文件变动触发，如果是 src 目录下的文件变动，则不做 restart
      async fileChange(e: {
        filename: string;
        eventType: string;
        isPrivate: boolean;
      }) {
        const { filename, eventType, isPrivate } = e;

        if (!isPrivate && (eventType === 'change' || eventType === 'unlink')) {
          const { closeServer } = await import('./utils/createServer');
          await closeServer();
          await restart(api.useHookRunners(), filename);
        }
      },

      async beforeRestart() {
        cleanRequireCache([require.resolve('./plugins/analyze')]);
      },

      async modifyFileSystemRoutes({ entrypoint, routes }) {
        nestedRoutes[entrypoint.entryName] = routes;

        return {
          entrypoint,
          routes,
        };
      },

      async beforeGenerateRoutes({ entrypoint, code }) {
        const { distDirectory } = api.useAppContext();

        await fs.outputJSON(
          path.resolve(distDirectory, NESTED_ROUTE_SPEC_FILE),
          nestedRoutes,
        );

        return {
          entrypoint,
          code,
        };
      },
    };
  },
});

export default appTools;
