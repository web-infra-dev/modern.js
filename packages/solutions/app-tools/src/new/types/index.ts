import type { DevToolData, RegisterBuildPlatformResult } from '@modern-js/core';
import type { AppContext } from '@modern-js/plugin-v2/types';
import type {
  CLIPluginAPI,
  PluginHookTap,
  TransformFunction,
} from '@modern-js/plugin-v2/types';
import type {
  Entrypoint,
  NestedRouteForCli,
  PageRoute,
  RouteLegacy,
  ServerPlugin,
  ServerRoute,
} from '@modern-js/types';
import type {
  AppToolsNormalizedConfig,
  AppToolsUserConfig,
} from '../../types/config';
import type { RuntimePlugin } from '../../types/hooks';
import type { Bundler } from '../../types/utils';
import type { getHookRunners } from '../compat/hooks';

export type BeforeConfigFn = () => Promise<void> | void;
export type AfterPrepareFn = () => Promise<void> | void;
export type InternalRuntimePluginsFn = TransformFunction<{
  entrypoint: Entrypoint;
  plugins: RuntimePlugin[];
}>;
export type InternalServerPluginsFn = TransformFunction<{
  plugins: ServerPlugin[];
}>;
export type CheckEntryPointFn = TransformFunction<{
  path: string;
  entry: false | string;
}>;
export type ModifyEntrypointsFn = TransformFunction<Entrypoint[]>;
export type ModifyFileSystemRoutesFn = TransformFunction<{
  entrypoint: Entrypoint;
  routes: RouteLegacy[] | (NestedRouteForCli | PageRoute)[];
}>;
export type ModifyServerRoutesFn = TransformFunction<{ routes: ServerRoute[] }>;
export type DeplpoyFn = () => Promise<void> | void;
export type GenerateEntryCodeFn = (
  entrypoints: Entrypoint[],
) => Promise<void> | void;
export type BeforeGenerateRoutesFn = TransformFunction<{
  entrypoint: Entrypoint;
  code: string;
}>;
export type BeforePrintInstructionsFn = TransformFunction<{
  instructions: string;
}>;
export type RegisterDevFn = () => Promise<DevToolData> | DevToolData;
export type RegisterBuildPlatformFn = () =>
  | Promise<RegisterBuildPlatformResult>
  | RegisterBuildPlatformResult;
export type AddRuntimeExportsFn = () => Promise<void> | void;

interface AppToolsExtendAPI<B extends Bundler = 'webpack'> {
  onBeforeConfig: PluginHookTap<BeforeConfigFn>;
  onAfterPrepare: PluginHookTap<AfterPrepareFn>;
  deploy: PluginHookTap<DeplpoyFn>;

  _internalRuntimePlugins: PluginHookTap<InternalRuntimePluginsFn>;
  _internalServerPlugins: PluginHookTap<InternalServerPluginsFn>;
  checkEntryPoint: PluginHookTap<CheckEntryPointFn>;
  modifyEntrypoints: PluginHookTap<ModifyEntrypointsFn>;
  modifyFileSystemRoutes: PluginHookTap<ModifyFileSystemRoutesFn>;
  modifyServerRoutes: PluginHookTap<ModifyServerRoutesFn>;

  generateEntryCode: PluginHookTap<GenerateEntryCodeFn>;
  onBeforeGenerateRoutes: PluginHookTap<BeforeGenerateRoutesFn>;
  /**
   * @deprecated
   */
  onBeforePrintInstructions: PluginHookTap<BeforePrintInstructionsFn>;
  /**
   * @deprecated
   */
  registerDev: PluginHookTap<RegisterDevFn>;
  /**
   * @deprecated
   */
  registerBuildPlatform: PluginHookTap<RegisterBuildPlatformFn>;
  /**
   * @deprecated
   */
  addRuntimeExports: PluginHookTap<AddRuntimeExportsFn>;

  /**
   * @deprecated use getAppContext instead
   */
  useAppContext: () => AppContext<
    AppToolsUserConfig<B>,
    AppToolsNormalizedConfig<AppToolsUserConfig<B>>
  >;
  /**
   * @deprecated use getConfig instead
   */
  useConfigContext: () => AppToolsUserConfig<B>;
  /**
   * @deprecated use getNormalizedConfig instead
   */
  useResolvedConfigContext: () => AppToolsNormalizedConfig<
    AppToolsUserConfig<B>
  >;
  /**
   * @deprecated use api.xx instead
   */
  useHookRunners: () => ReturnType<typeof getHookRunners>;
}
export interface AppTools<B extends Bundler = 'webpack'>
  extends CLIPluginAPI<
      AppToolsUserConfig<B>,
      AppToolsNormalizedConfig<AppToolsUserConfig<B>>
    >,
    AppToolsExtendAPI<B> {}

export type AppToolsExtendAPIName<B extends Bundler = 'webpack'> =
  keyof AppToolsExtendAPI<B> & string;
