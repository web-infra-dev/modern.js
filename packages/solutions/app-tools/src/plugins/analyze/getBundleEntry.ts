import path from 'path';
import {
  ensureAbsolutePath,
  fs,
  findExists,
  MAIN_ENTRY_NAME,
  JS_EXTENSIONS,
} from '@modern-js/utils';
import type { Entrypoint } from '@modern-js/types';
import { CliHooksRunner } from '@modern-js/core';
import type { AppNormalizedConfig, AppTools, IAppContext } from '../../types';
import { getFileSystemEntry } from './getFileSystemEntry';
import { isSubDirOrEqual } from './utils';

const ensureExtensions = (file: string) => {
  if (!path.extname(file)) {
    return findExists(JS_EXTENSIONS.map(ext => `${file}${ext}`)) || file;
  }
  return file;
};

/**
 *
 * Lightweight method for whether it is a directory.
 */
const isDirectory = (file: string) => !path.extname(file);

const ifAlreadyExists = (
  entrypoints: Entrypoint[],
  checked: Entrypoint,
): boolean =>
  entrypoints.some(entrypoint => {
    if (
      ensureExtensions(entrypoint.entry) === ensureExtensions(checked.entry)
    ) {
      // reset entryName
      checked.entryName = entrypoint.entryName;
      return true;
    }
    // filesystem routes entrypoint conflict with normal entrypoint.
    if (
      isSubDirOrEqual(entrypoint.entry, checked.entry) ||
      isSubDirOrEqual(checked.entry, entrypoint.entry)
    ) {
      throw new Error(
        `Entry configuration conflicts\n Your configuration: ${checked.entry}.\n Default entrypoint: ${entrypoint.entry}\n Please reset source.entries or set source.disableDefaultEntries to disable the default entry rules.`,
      );
    }

    return false;
  });

export const getBundleEntry = async (
  hookRunners: CliHooksRunner<AppTools<'shared'>>,
  appContext: IAppContext,
  config: AppNormalizedConfig<'shared'>,
) => {
  const { appDirectory, packageName } = appContext;
  const { disableDefaultEntries, entries, entriesDir, mainEntryName } =
    config.source;

  const defaults = disableDefaultEntries
    ? []
    : await getFileSystemEntry(hookRunners, appContext, config);

  // merge entrypoints from user config with directory convention.
  if (entries) {
    Object.keys(entries).forEach(name => {
      const value = entries[name];
      const entryName = typeof value === 'string' ? value : value.entry;
      const isAutoMount =
        typeof value === 'string' ? true : !value.disableMount;
      const entrypoint: Entrypoint = {
        entryName: name,
        isMainEntry: false,
        entry: ensureAbsolutePath(appDirectory, entryName),
        absoluteEntryDir: isDirectory(
          ensureAbsolutePath(appDirectory, entryName),
        )
          ? ensureAbsolutePath(appDirectory, entryName)
          : path.dirname(ensureAbsolutePath(appDirectory, entryName)),
        isAutoMount,
        customBootstrap:
          typeof value === 'string'
            ? false
            : value.customBootstrap &&
              ensureAbsolutePath(appDirectory, value.customBootstrap),
        fileSystemRoutes: fs
          .statSync(ensureAbsolutePath(appDirectory, entryName))
          .isDirectory()
          ? {}
          : undefined,
      };

      if (!ifAlreadyExists(defaults, entrypoint)) {
        defaults.push(entrypoint);
      }
    });
  }

  if (!disableDefaultEntries) {
    // find main entry point which server route is '/'.
    const entriesDirAbs = ensureAbsolutePath(appDirectory, entriesDir || '');
    const found = defaults.find(
      ({ entryName, entry, nestedRoutesEntry = '' }) =>
        entryName === packageName ||
        path.dirname(entry) === entriesDirAbs ||
        path.dirname(nestedRoutesEntry) === entriesDirAbs,
    );
    if (found) {
      found.entryName = mainEntryName || MAIN_ENTRY_NAME;
      found.isMainEntry = true;
    }
  }
  return defaults;
};
