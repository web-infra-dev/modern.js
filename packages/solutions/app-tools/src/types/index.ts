import type { NormalizedConfig, UserConfig } from '@modern-js/core';
import type { AppToolsUserConfig, AppToolsNormalizedConfig } from './config';
import type { AppToolsHooks } from './hooks';
import type {
  AppToolsLegacyUserConfig,
  AppToolsLegacyNormalizedConfig,
} from './legacyConfig';
import { Bundler } from './utils';

export * from './hooks';
export * from './config';
export * from './legacyConfig';
export type { webpack, Rspack } from '@modern-js/uni-builder';
export type { Bundler } from './utils';
export type {
  ServerUserConfig,
  ServerNormalizedConfig,
  BffUserConfig,
  BffNormalizedConfig,
  SSR,
  SSRByEntries,
} from '@modern-js/server-core';
export type {
  IAppContext,
  PluginAPI,
  CliPlugin,
  NormalizedConfig,
  UserConfig,
} from '@modern-js/core';

export type AppTools<B extends Bundler = 'webpack'> = {
  hooks: AppToolsHooks<B>;
  userConfig: AppToolsUserConfig<B>;
  normalizedConfig: AppToolsNormalizedConfig<AppToolsUserConfig<'shared'>>;
};

export type LegacyAppTools = {
  hooks: AppToolsHooks;
  userConfig: AppToolsLegacyUserConfig;
  normalizedConfig: AppToolsLegacyNormalizedConfig;
};

export type AppNormalizedConfig<B extends Bundler = 'webpack'> =
  NormalizedConfig<AppTools<B>>;
export type AppLegacyNormalizedConfig = NormalizedConfig<LegacyAppTools>;

export type AppUserConfig<B extends Bundler = 'webpack'> = UserConfig<
  AppTools<B>
>;
export type AppLegacyUserConfig = UserConfig<LegacyAppTools>;
