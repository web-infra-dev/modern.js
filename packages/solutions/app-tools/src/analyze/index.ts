import * as path from 'path';
import { createDebugger, fs, isApiOnly } from '@modern-js/utils';
import type { CliPlugin } from '@modern-js/core';
import { cloneDeep } from '@modern-js/utils/lodash';
import { createBuilderForModern } from '../builder';
import { printInstructions } from '../utils/printInstructions';
import { generateRoutes } from '../utils/routes';
import { emitResolvedConfig } from '../utils/config';
import { getCommand } from '../utils/commands';
import { AppTools } from '../types';
import { initialNormalizedConfig } from '../config';
import { isRouteComponentFile } from './utils';
import { loaderBuilder, serverLoaderBuilder } from './Builder';

const debug = createDebugger('plugin-analyze');

export default (): CliPlugin<AppTools> => ({
  name: '@modern-js/plugin-analyze',

  setup: api => {
    let pagesDir: string[] = [];
    let originEntrypoints: any[] = [];

    return {
      async prepare() {
        let appContext = api.useAppContext();
        const resolvedConfig = api.useResolvedConfigContext();
        const hookRunners = api.useHookRunners();

        try {
          fs.emptydirSync(appContext.internalDirectory);
        } catch {
          // FIXME:
        }

        const apiOnly = await isApiOnly(
          appContext.appDirectory,
          resolvedConfig.source?.entriesDir,
        );
        await hookRunners.addRuntimeExports();

        if (apiOnly) {
          const { routes } = await hookRunners.modifyServerRoutes({
            routes: [],
          });

          debug(`server routes: %o`, routes);

          appContext = {
            ...appContext,
            apiOnly,
            serverRoutes: routes,
          };
          api.setAppContext(appContext);
          return;
        }

        const [
          { getBundleEntry },
          { getServerRoutes },
          { generateCode },
          { getHtmlTemplate },
        ] = await Promise.all([
          import('./getBundleEntry'),
          import('./getServerRoutes'),
          import('./generateCode'),
          import('./getHtmlTemplate'),
        ]);

        const entrypoints = getBundleEntry(appContext, resolvedConfig);
        const defaultChecked = entrypoints.map(point => point.entryName);

        debug(`entrypoints: %o`, entrypoints);

        const initialRoutes = getServerRoutes(entrypoints, {
          appContext,
          config: resolvedConfig,
        });

        const { routes } = await hookRunners.modifyServerRoutes({
          routes: initialRoutes,
        });

        debug(`server routes: %o`, routes);

        appContext = {
          ...appContext,
          entrypoints,
          serverRoutes: routes,
        };
        api.setAppContext(appContext);

        const nestedRouteEntries = entrypoints
          .map(point => point.nestedRoutesEntry)
          .filter(Boolean) as string[];

        pagesDir = entrypoints
          .map(point => point.entry)
          .filter(Boolean)
          .concat(nestedRouteEntries);

        originEntrypoints = cloneDeep(entrypoints);

        await generateCode(appContext, resolvedConfig, entrypoints, api);

        const htmlTemplates = await getHtmlTemplate(entrypoints, api, {
          appContext,
          config: resolvedConfig,
        });

        debug(`html templates: %o`, htmlTemplates);

        await hookRunners.addDefineTypes();

        debug(`add Define Types`);

        appContext = {
          ...appContext,
          entrypoints,
          checkedEntries: defaultChecked,
          apiOnly,
          serverRoutes: routes,
          htmlTemplates,
        };
        api.setAppContext(appContext);

        const command = getCommand();
        const buildCommands = ['dev', 'build', 'inspect', 'deploy'];
        if (buildCommands.includes(command)) {
          const normalizedConfig = api.useResolvedConfigContext();
          const builder = await createBuilderForModern({
            normalizedConfig: normalizedConfig as any,
            appContext,
            compatPluginConfig: {
              async onBeforeBuild({ bundlerConfigs }) {
                const hookRunners = api.useHookRunners();
                await generateRoutes(appContext);
                await hookRunners.beforeBuild({ bundlerConfigs });
              },

              async onAfterBuild({ stats }) {
                const hookRunners = api.useHookRunners();
                await hookRunners.afterBuild({ stats });
                await emitResolvedConfig(
                  appContext.appDirectory,
                  normalizedConfig,
                );
              },

              async onDevCompileDone({ isFirstCompile }) {
                const hookRunners = api.useHookRunners();
                if (process.stdout.isTTY || isFirstCompile) {
                  hookRunners.afterDev();

                  if (isFirstCompile) {
                    printInstructions(
                      hookRunners,
                      appContext,
                      normalizedConfig,
                    );
                  }
                }
              },

              async onBeforeCreateCompiler({ bundlerConfigs }) {
                const hookRunners = api.useHookRunners();
                // run modernjs framework `beforeCreateCompiler` hook
                await hookRunners.beforeCreateCompiler({
                  bundlerConfigs,
                });
              },

              async onAfterCreateCompiler({ compiler }) {
                const hookRunners = api.useHookRunners();
                // run modernjs framework afterCreateCompiler hooks
                await hookRunners.afterCreateCompiler({ compiler });
              },
            },
          });

          builder.addPlugins(resolvedConfig.builderPlugins);

          appContext = {
            ...appContext,
            builder,
          };
          api.setAppContext(appContext);
        }
      },
      watchFiles() {
        return pagesDir;
      },

      resolvedConfig({ resolved }) {
        const appContext = api.useAppContext();
        const config = initialNormalizedConfig(resolved, appContext);
        return {
          resolved: config,
        };
      },

      async beforeRestart() {
        serverLoaderBuilder.stop();
        loaderBuilder.stop();
      },

      async fileChange(e) {
        const appContext = api.useAppContext();
        const { appDirectory } = appContext;
        const { filename, eventType } = e;
        const isPageFile = (name: string) =>
          pagesDir.some(pageDir => name.includes(pageDir));

        const absoluteFilePath = path.resolve(appDirectory, filename);
        const isRouteComponent =
          isPageFile(absoluteFilePath) &&
          isRouteComponentFile(absoluteFilePath);

        if (
          isRouteComponent &&
          (eventType === 'add' || eventType === 'unlink')
        ) {
          const resolvedConfig = api.useResolvedConfigContext();
          const { generateCode } = await import('./generateCode');
          const entrypoints = cloneDeep(originEntrypoints);
          generateCode(appContext, resolvedConfig, entrypoints, api);
        }
      },
    };
  },
});
