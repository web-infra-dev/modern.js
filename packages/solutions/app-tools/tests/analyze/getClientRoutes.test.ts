import path from 'path';
import type { IAppContext } from '@modern-js/core';
import { getBundleEntry } from '../../src/analyze/getBundleEntry';
import {
  getClientRoutes,
  getClientRoutesLegacy,
} from '../../src/analyze/getClientRoutes';
import { AppNormalizedConfig } from '../../src/types';

type GetClientRoutesFunc =
  | typeof getClientRoutesLegacy
  | typeof getClientRoutes;

const prepareEnv = (fixturePath: string) => {
  const appContext = {
    appDirectory: fixturePath,
    srcDirectory: path.join(fixturePath, 'src'),
    internalDirectory: path.resolve(fixturePath, './node_modules/.modern-js'),
    internalDirAlias: '@_modern_js_internal',
    internalSrcAlias: '@_modern_js_src',
  };

  const config = { source: { entriesDir: './src' } };

  return {
    appContext,
    config,
  };
};

describe('getClientRoutesLegacy', () => {
  test('basic usage', () => basicUsage(getClientRoutesLegacy));

  test('nested routes with `_app.tsx` and `_layout.tsx`', () =>
    supportLayout(getClientRoutesLegacy));
});

describe('getClientRoutes', () => {
  test('basic usage', () => basicUsage(getClientRoutes));

  test('nested routes with `_app.tsx` and `_layout.tsx`', () =>
    supportLayout(getClientRoutes));
});

function basicUsage(getClientRoutes: GetClientRoutesFunc) {
  const fixturePath = path.resolve(
    __dirname,
    './fixtures/entries/file-system-routes',
  );

  const { appContext, config } = prepareEnv(fixturePath);

  const entries = getBundleEntry(
    appContext as IAppContext,
    config as AppNormalizedConfig<'shared'>,
  );

  let routes;

  for (const entrypoint of entries) {
    routes = getClientRoutes({
      entrypoint,
      srcDirectory: appContext.srcDirectory,
      srcAlias: appContext.internalSrcAlias,
      internalDirectory: appContext.internalDirectory,
      internalDirAlias: appContext.internalDirAlias,
    });
  }

  expect(routes).toMatchSnapshot();
}

function supportLayout(getClientRoutes: GetClientRoutesFunc) {
  const fixturePath = path.resolve(
    __dirname,
    './fixtures/entries/file-system-routes-nested',
  );

  const { appContext, config } = prepareEnv(fixturePath);

  const entries = getBundleEntry(
    appContext as IAppContext,
    config as AppNormalizedConfig<'shared'>,
  );

  let routes;

  for (const entrypoint of entries) {
    routes = getClientRoutes({
      entrypoint,
      srcDirectory: appContext.srcDirectory,
      srcAlias: appContext.internalSrcAlias,
      internalDirectory: appContext.internalDirectory,
      internalDirAlias: appContext.internalDirAlias,
    });
  }

  expect(routes).toMatchSnapshot();
}
