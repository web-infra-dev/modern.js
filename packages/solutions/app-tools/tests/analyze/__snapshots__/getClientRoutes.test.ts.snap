// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getClientRoutes basic usage 1`] = `
Array [
  Object {
    "_component": "@_modern_js_src/pages/index.ts",
    "component": "Comp_pages_index_ts",
    "parent": undefined,
    "path": "/",
    "type": "page",
  },
  Object {
    "_component": "@_modern_js_src/pages/info.ts",
    "component": "Comp_pages_info_ts",
    "parent": undefined,
    "path": "/info",
    "type": "page",
  },
  Object {
    "_component": "@_modern_js_src/pages/about.ts",
    "component": "Comp_pages_about_ts",
    "parent": undefined,
    "path": "/about",
    "type": "page",
  },
]
`;

exports[`getClientRoutes nested routes with \`_app.tsx\` and \`_layout.tsx\` 1`] = `
Array [
  Object {
    "_component": "@_modern_js_src/pages/index.ts",
    "component": "Comp_pages_index_ts",
    "parent": undefined,
    "path": "/",
    "type": "page",
  },
  Object {
    "_component": "@_modern_js_internal/main/internal_components/L_a_Comp_pages_a_index_ts.jsx",
    "component": "L_a_Comp_pages_a_index_ts",
    "parent": undefined,
    "path": "/a",
    "type": "page",
  },
  Object {
    "_component": "@_modern_js_internal/main/internal_components/L_a_Comp_pages_a_list_index_ts.jsx",
    "component": "L_a_Comp_pages_a_list_index_ts",
    "parent": undefined,
    "path": "/a/list",
    "type": "page",
  },
  Object {
    "_component": "@_modern_js_internal/main/internal_components/L_a_Comp_pages_a_about_index_ts.jsx",
    "component": "L_a_Comp_pages_a_about_index_ts",
    "parent": undefined,
    "path": "/a/about",
    "type": "page",
  },
  Object {
    "_component": "@_modern_js_internal/main/internal_components/L_a_Comp_pages_a_list__id__index_ts.jsx",
    "component": "L_a_Comp_pages_a_list__id__index_ts",
    "parent": undefined,
    "path": "/a/list/:id",
    "type": "page",
  },
]
`;

exports[`getClientRoutesLegacy basic usage 1`] = `
Array [
  Object {
    "_component": "@_modern_js_src/pages/index.ts",
    "component": "Comp_pages_index_ts",
    "exact": true,
    "parent": undefined,
    "path": "/",
  },
  Object {
    "_component": "@_modern_js_src/pages/info.ts",
    "component": "Comp_pages_info_ts",
    "exact": true,
    "parent": undefined,
    "path": "/info",
  },
  Object {
    "_component": "@_modern_js_src/pages/about.ts",
    "component": "Comp_pages_about_ts",
    "exact": true,
    "parent": undefined,
    "path": "/about",
  },
]
`;

exports[`getClientRoutesLegacy nested routes with \`_app.tsx\` and \`_layout.tsx\` 1`] = `
Array [
  Object {
    "_component": "@_modern_js_src/pages/index.ts",
    "component": "Comp_pages_index_ts",
    "exact": true,
    "parent": undefined,
    "path": "/",
  },
  Object {
    "_component": "@_modern_js_internal/main/internal_components/L_a_Comp_pages_a_index_ts.jsx",
    "component": "L_a_Comp_pages_a_index_ts",
    "exact": true,
    "parent": undefined,
    "path": "/a",
  },
  Object {
    "_component": "@_modern_js_internal/main/internal_components/L_a_Comp_pages_a_list_index_ts.jsx",
    "component": "L_a_Comp_pages_a_list_index_ts",
    "exact": true,
    "parent": undefined,
    "path": "/a/list",
  },
  Object {
    "_component": "@_modern_js_internal/main/internal_components/L_a_Comp_pages_a_about_index_ts.jsx",
    "component": "L_a_Comp_pages_a_about_index_ts",
    "exact": true,
    "parent": undefined,
    "path": "/a/about",
  },
  Object {
    "_component": "@_modern_js_internal/main/internal_components/L_a_Comp_pages_a_list__id__index_ts.jsx",
    "component": "L_a_Comp_pages_a_list__id__index_ts",
    "exact": true,
    "parent": undefined,
    "path": "/a/list/:id",
  },
]
`;
